### PR TITLE
update release note generation

### PR DESCRIPTION
**Pull request by radTuti**
## Description

- ensure milestone is closed before generating release notes
- ensure there are no open PRs in the milestone

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Review by danudey — CHANGES_REQUESTED** [review]

File: hack/generate_release_notes.py [hunk 1]
```diff
@@ -47,27 +47,37 @@ def issues_in_milestone() -> list:
...
+        issue for issue in milestone_issues if issue.as_pull_request().merged
+    ]
+    # If there are open issues, raise an error.
+    if len(open_issues) > 0:
+        raise ReleaseNoteError(
+            f"{len(open_issues)} PRs are still open, remove from milestone... skipping"
+        )
+    # Since we only want merged PRs, raise an error if there are none.
+    if len(merged_issues) == 0:
+        raise ReleaseNoteError(f"no issues found for milestone {m.title}")
```
[the remark this answers is not in the document]
Comment (radTuti) [reply hack/generate_release_notes.py:79]: In that case, we should not be using this tooling to generate release notes.

**Review by danudey — APPROVED** [review]

File: hack/generate_release_notes.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danudey) [reply hack/generate_release_notes.py:79]: It still generates the outline for the release notes, so it's still valuable to run.

File: RELEASING.md [hunk 2]
```diff
@@ -95,64 +101,23 @@ Go to the PR created and:
 
 1. Create a git tag for the new commit on the release branch and push it:
 
-```
-git tag v1.30.3
-git push --tags
-```
+    ```sh
+    git tag v1.30.3
+    git push --tags
```
Comment (tmjd) [inline RELEASING.md:106]: I'd suggest pushing only the newly created tag. If a user has other tags locally we probably don't want them all pushed.

Comment (caseydavenport) [reply RELEASING.md:106]: Yep, 100% this.

File: RELEASING.md [hunk 3]
```diff
@@ -95,64 +101,23 @@ Go to the PR created and:
...
-
-   ```
-   make release-notes VERSION=<TAG> GITHUB_TOKEN=<access-token>
-   ```
+1. Go to [releases](https://github.com/tigera/operator/releases) and edit the draft release for the release tag
 
-1. Go to https://github.com/tigera/operator/releases and edit the release tag to include the generated release notes, and update the title.
+1. Publish the release.
 
-1. Close the milestone for this release. https://github.com/tigera/operator/milestones
```
Comment (tmjd) [inline RELEASING.md:117]: This still must be done AFAICT, don't we need to move this before doing the 'Rerun' step above or else the Rerun won't create the release draft?

**Comment by tmjd** [comment]
Seems like these checks on the milestone and PRs in the milestone should perhaps be done as part of the `release-prep` or there should be a check before doing the "Rerun" in semaphore.

File: RELEASING.md (hunk 3 above)
Comment (radTuti) [reply RELEASING.md:117]: The release notes now require that the milestone is closed first before generating the draft release with release notes.

**Comment by radTuti** [comment]
> Seems like these checks on the milestone and PRs in the milestone should perhaps be done as part of the `release-prep` also or there should be a check before doing the "Rerun" in semaphore.

I have moved the checking milestones and PRs in milestones to "Preparing for the release" section in the releasing docs. The check in the release notes are additional sanity check.
I have also added to the `release-prep` task to handle checking the milestone

File: Makefile [hunk 4]
```diff
@@ -565,7 +565,19 @@ ifdef LOCAL_BUILD
...
+handle-milestone: hack/bin/gh var-require-all-VERSION-GITHUB_TOKEN
+	@gh extension install valeriobelli/gh-milestone
+	@echo "Checking milestone $(VERSION) exists"
+	@gh milestone list --query $(VERSION) --repo $(REPO) --state all --json title | \
+		jq 'if .[0] == [] then error("milestone for release does not exist") else empty end'
+	@echo "Checking $(VERSION) milestone has no open PRs"
+	@gh search prs --milestone $(VERSION) --state open --json id | \
+		jq 'if . == [] then empty else error("there are open PRs in milestone") end'
+	$(eval MILESTONE_NUMBER := $(shell gh milestone list --query $(VERSION) --repo $(REPO) --state closed --json number --jq '.[0].number'))
+	@echo "Closing $(VERSION) milestone #$(MILESTONE_NUMBER)"
```
Comment (tmjd) [inline Makefile:577]: You've already got in the readme to close the Milestone, I don't think we should do that here also, I think here we should just check that the milestone is closed.

Comment (radTuti) [reply Makefile]: updated to check that the milestone is closed

**Review by tmjd — APPROVED** [review]
LGTM

**Pull request merged**